### ISSUE TITLE
DoRA training not taking dropout or alpha into account

### ISSUE THREAD

**Issue by BenjaminBossan** [AnswerDotAI/fsdp_qlora#68]
I think there is a bug in the DoRA implementation as it takes neither `lora_dropout` nor `lora_alpha` into account. These arguments are passed as `*args` to the `__init__` call of the DoRA layers but subsequently ignored inside of [dora.py](https://github.com/AnswerDotAI/fsdp_qlora/blob/05ed9f2a60f96a0795cb082bceab70a9b19fd213/scripts/dora.py). This can be easily missed as the DoRA paper does not include them in their equations, but they are mentioned elsewhere in the paper and should be applied the same as in the LoRA implementation.

Also note that `lora_dropout` is only applied to the LoRA/DoRA output, not the [base model output](https://github.com/AnswerDotAI/fsdp_qlora/blob/05ed9f2a60f96a0795cb082bceab70a9b19fd213/scripts/dora.py#L92), which I believe has an impact on [these lines](https://github.com/AnswerDotAI/fsdp_qlora/blob/05ed9f2a60f96a0795cb082bceab70a9b19fd213/scripts/dora.py#L111-L113), as they currently assume that the same `x` is used for the base layer and the DoRA part.

**Comment by devtyagi3909** [comment]
yeah this is a real issue. `DORALayer` `__init__` eats `*args` and `**kwargs` but never extracts or applies `lora_dropout` and `lora_alpha`. the missing `lora_alpha` means the DoRA forward pass is missing the scaling factor `alpha / rank` entirely. we need to apply `scaling * (self.lora_B(self.lora_A(dropout(x))))` in the forward pass, and more importantly, bake that `scaling` into the column norm computation: `(frozen_weight + scaling * self.lora_B.weight @ self.lora_A.weight).norm(p=2, dim=1)`. if that sounds right to you, i can prototype the fix.
